### ISSUE TITLE
Middleware and CurrentAttributes 

### ISSUE THREAD

**Issue by birdwell** [mperham/sidekiq#4568]
Ruby version: 2.6.1
Sidekiq / Pro / Enterprise version(s): Pro

Please include your initializer and any error message with the full backtrace.
In extras, we are using middleware to set Current.attributes from "current_context". Current isn't getting set correctly for one of our jobs even though current_context has the correct info. Then after a retry, current_context is empty so all retries will fail.

```
Current.attributes = job['current_context'].map { |k,v| [k, v.instance_of?(Hash) ? RecursiveOpenStruct.new(v) : v] }.to_h.symbolize_keys
```

```
require_all 'lib/sidekiq/middleware/'
require 'sidekiq-status'

unless Rails.env.development? && !ENV['DEBUG_SIDEKIQ_DEV']

  Sidekiq::Client.reliable_push! unless Rails.env.test?

  SidekiqUniqueJobs.configure do |config|
    config.enabled = !Rails.env.test?
  end

  Sidekiq.configure_client do |config|
    config.redis = { url: "redis", size: 2, id: nil }

    config.client_middleware do |chain|
      chain.add Sidekiq::Middleware::Client::CurrentContext
    end

    Sidekiq::Status.configure_client_middleware config, expiration: 30.minutes
  end

  Sidekiq.configure_server do |config|
    config.redis = { url: "redis", size: 2,  id: nil }
    config.logger.formatter = Logging::JsonFormatter.new
    config.super_fetch!

    config.client_middleware do |chain|
      chain.add Sidekiq::Middleware::Client::CurrentContext
    end
    config.server_middleware do |chain|
      chain.add Sidekiq::Middleware::Server::CurrentContext
    end

    Sidekiq::Status.configure_server_middleware config, expiration: 30.minutes
    Sidekiq::Status.configure_client_middleware config, expiration: 30.minutes
  end

end
```

Are you using an old version? No
Have you checked the changelogs to see if your issue has been fixed in a later version? 5.0.1, latest

https://github.com/mperham/sidekiq/blob/master/Changes.md
https://github.com/mperham/sidekiq/blob/master/Pro-Changes.md
https://github.com/mperham/sidekiq/blob/master/Ent-Changes.md

**Comment by mperham** [comment]
You've not attached the most important bit: your CurrentContext middleware code.

**Comment by birdwell** [comment]
It's there the guts. Here is the rest.

```
module Sidekiq
  module Middleware
    module Server
      class CurrentContext
        def call(worker, job, queue)
          Current.attributes = job['current_context'].map { |k,v| [k, v.instance_of?(Hash) ? RecursiveOpenStruct.new(v) : v] }.to_h.symbolize_keys
          yield
        end
      end
    end
  end
end
```

**Comment by mperham** [comment]
Tip: use the hammer operator.

```ruby
job['current_context'] ||= ....
```

Otherwise, Pushing the job to Redis to retry will mutate the current_content.

**Comment by birdwell** [comment]
I will add that right away. Is there anything else that stands out? I can provide more insight or close  the ticket? If it comes down to what we are doing with CurrentAttributes that is the issue then it's not a sidekiq conversation. I wanted to make sure we weren't overlooking at how CurrentAttributes gets loaded in when a job is executed. Thanks @mperham I'm new to all the ruby stuff so I'm just trying to dig in and learn.

**Comment by mperham** [comment]
That's all I've got. Close it if you find it behaving better, otherwise let us know any new issues.

**Comment by birdwell** [comment]
Thanks. 😄

**Comment by khun84** [comment]
Some word of caution on using `ActiveSupport::CurrentAttributes` in sidekiq environment.

Even if one had setup the `Current` attributes in the middleware, but those attributes will be reset sometime later during before and after the `yield` block. 

This behaviour is coming from the callbacks that was setup by `ActiveSupport::CurrentAttributes` [initializer](https://github.com/rails/rails/blob/645239817d15c36ff008ccb5b6c8887e3b2c8f4b/activesupport/lib/active_support/railtie.rb#L46) which it will reset all attributes on before and after the Rails app executor runs. Sidekiq uses the same app executor to wrap the execution of worker code.

In short, `Current` is not so reliable in Sidekiq environment as any breaking changes on the library or sidekiq on callbacks and execution wrapping would break your expectation and it may not be obvious to you. I personally bumped into this situation recently and have changed the global store  implementation to `RequestStore` in sidekiq env.

**Comment by jordan-brough** [comment]
@khun84 can you explain more about the behavior you were seeing?
Was this in a production environment or in a development environment?

> Even if one had setup the Current attributes in the middleware, but those attributes will be reset sometime later during before and after the yield block.

You're saying that if you use `ActiveSupport::CurrentAttributes` with Sidekiq, the attributes will be reset at random(?) times when processing jobs?

> This behaviour is coming from the callbacks that was setup by ActiveSupport::CurrentAttributes [initializer](https://github.com/rails/rails/blob/645239817d15c36ff008ccb5b6c8887e3b2c8f4b/activesupport/lib/active_support/railtie.rb#L46)

The code you linked to appears to be related to code reloading that happens in a development environment, but I assume wouldn't happen in a production environment. Are you seeing something else?
